NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,531 @@
+        # create an output folder and specify the file path if create_tree=True
+        if create_tree:
+            os.makedirs('output', exist_ok=True)
+            os.makedirs('output/trees', exist_ok=True)
+            tree_dir = './output/trees/'
+        else:
+            tree_dir = None
+    
+        # create an empty list to store all detected quotes
+        all_quotes = []
+        
+        # go through all the texts and start extracting quotes
+        for row in text_df.itertuples():
+            doc_id = row.Index
+            doc = row.spacy_text
+            
+            try:        
+                # extract the quotes
+                quotes = extract_quotes(doc_id=doc_id, doc=doc, 
+                                        write_tree=create_tree, 
+                                        tree_dir=tree_dir)
+                
+                # extract the included named entities
+                speaks, qts = [quote['speaker'] for quote in quotes], [quote['quote'] for quote in quotes]        
+                speak_ents = [[(str(ent), ent.label_) for ent in doc.ents if (str(ent) in speak) & (ent.label_ in inc_ent)] for speak in speaks]        
+                quote_ents = [[(str(ent), ent.label_) for ent in doc.ents if (str(ent) in qt) & (ent.label_ in inc_ent)] for qt in qts]
+        
+                # add text_id, quote_id and named entities to each quote
+                for n, quote in enumerate(quotes):
+                    quote['text_id'] = doc_id
+                    quote['quote_id'] = str(n)
+                    quote['speaker_entities'] = list(set(speak_ents[n]))
+                    quote['quote_entities'] = list(set(quote_ents[n]))
+                    
+                # store them in all_quotes
+                all_quotes.extend(quotes)
+                    
+            except:
+                # this will provide some information in the case of an error
+                self.app_logger.exception("message")
+                traceback.print_exception()
+                
+        # convert the outcome into a pandas dataframe
+        quotes_df = pd.DataFrame.from_dict(all_quotes)
+        
+        # convert the string format quote spans in the index columns to a tuple of integers
+        for column in quotes_df.columns:
+            if column.endswith('_index'):
+                quotes_df[column].replace('','(0,0)', inplace=True)
+                quotes_df[column] = quotes_df[column].apply(eval)
+        
+        # re-arrange the columns
+        new_index = ['text_id', 'quote_id', 'quote', 'quote_index', 'quote_entities', 
+                     'speaker', 'speaker_index', 'speaker_entities',
+                     'verb', 'verb_index', 'quote_token_count', 'quote_type', 'is_floating_quote']
+        quotes_df = quotes_df.reindex(columns=new_index)
+                
+        return quotes_df
+    
+    
+    def show_quotes(self, text_df, quotes_df, text_id, show_what, inc_ent, save_to_html=False):
```

Comment:
We can/should make text_df, quotes_df part of the quotation_tool class? As these are critical data for the tool to work, after initialisation we can just use self.text_df, self.quotes_df rather than passing as arguments?

Reply:
self.text_df and self.quotes_df have been added and code modified accordingly